Improved efficiency of overlap_sort

caseyflex opened the pull request:
Uses the faster outer_dot to improve efficiency of overlap_sort. Sometimes, sorting modes can be a bottleneck in mode solving, and this should prevent that from occuring.

Based off of https://github.com/flexcompute/tidy3d/pull/1464

caseyflex commented:
> @caseyflex should we close this (or make a draft) in view of #1464 ?

no sorry -- we should close this one: https://github.com/flexcompute/tidy3d/pull/1463

This PR uses the revised outer_dot PR https://github.com/flexcompute/tidy3d/pull/1464 to improve overlap_sort.

tylerflex commented:
ok I see, sorry for the confusion

momchil-flex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1126,6 +1127,18 @@ def _isel(self, **isel_kwargs):
+        ``self.monitor.mode_spec`` and ``self.monitor.freqs`` will no longer be matching the
+        newly created data."""
+
+        update_dict = dict(self._grid_correction_dict, **self.field_components)
```
_grid_correction_dict is a property so it shouldn't be added here at all?

caseyflex replied:
I guess this is adding `grid_primal_correction` and `grid_dual_correction` to the update list. I copied this from `_isel`, but it is probably not needed as these fields keep their old values...

caseyflex replied:
oh actually, it does do something -- it updates the coordinates on those two fields as well as the field_components. This may actually be needed so best to leave it.

momchil-flex replied:
Ah, got it.

momchil-flex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1081,6 +1081,7 @@ def overlap_sort(
             for freq_id in range(f0_ind + step, last_ind, step):
                 # Get next frequency to sort
                 data_to_sort = self._isel(f=[freq_id])
+                data_to_sort = data_to_sort._assign_coords(f=[self.monitor.freqs[f0_ind]])
```
I am confused, why is this needed? For a regular DataArray, isel-ing already redefines the coordinates, too. Is something not working right in our `_isel` wrapper?

![image](https://github.com/flexcompute/tidy3d/assets/92756559/9cb10443-416b-42f1-8446-3258332e55d2)

caseyflex replied:
It's a bit of a hack -- I want to isel `f=[freq_id]`, which is the next frequency to sort. But, I want to compute outer_dot against the base frequency `f0`. But because of how outer_dot works, it only keeps frequencies which are common between the two datasets. So, here, I isel `f=[freq_id]`, then I assign coords to `f0`, so that outer_dot will think they are at the same frequency and know to compare them.

momchil-flex replied:
Ohhhhh. Funny. One thing we could do also do is to allow `outer_dot` to work without a frequency dimension, and just pop it? But, eh, I think this is fine.

momchil-flex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -1147,19 +1160,20 @@ def _find_ordering_one_freq(
-            amps_reduced[:, i] = data_template_reduced.dot(one_mode).data.ravel()
+            # amps_reduced[:, i] = data_template_reduced.dot(one_mode).data.ravel()
             if self.monitor.store_fields_direction == "-":
                 amps_reduced[:, i] *= -1
```
Seems like the whole loop can be removed then since the `-1` factor can be applied to all of `amps_reduced`.

momchil-flex commented:
I don't know if @weiliangjin2021 still wants to look at this, but it should be good to merge after a rebase. The conflicts are probably in the changelog. You could squash the commits too.

caseyflex commented:
> I don't know if @weiliangjin2021 still wants to look at this, but it should be good to merge after a rebase. The conflicts are probably in the changelog. You could squash the commits too.

Rebased and squashed. Merge when you're ready. thanks!

weiliangjin2021 reviewed (approved):
I didn't take a very careful look. From what I understand, it all look good.
